libraries/chain/chain_controller.cpp

```diff
@@ -557,6 +548,17 @@ void chain_controller::_apply_block(const signed_block& next_block, uint32_t ski
    });
 }
 
+static void validate_shard_locks(const vector<shard_lock>& locks, const string& tag) {
+   if (locks.size() < 2) {
+      return;
+   }
+
+   for (auto cur = locks.begin() + 1; cur < locks.end(); cur++) {
+      auto prev = cur - 1;
+      FC_ASSERT(*prev != *cur, "${tag} lock \"${a}::${s}\" is not unique", ("tag",tag)("a",cur->account)("s",cur->scope));
+      FC_ASSERT(*prev < *cur,  "${tag} locks are not sorted", ("tag",tag));
+   }
```

Comment:
I see `FC_ASSERT` here and `EOS_ASSERT` elsewhere.  They do have slightly diff behavior. Could you give us guidance on when to use each?

Reply:
I think we should probably try to use EOS_ASSERT for anything which may trip as part of validating the block chain.   Its debatable, but code paths that are only exercised when a producer is producing a block don't benefit as much from EOS_ASSERT.  I have changed many of my FC_ASSERTs to EOS_ASSERTs